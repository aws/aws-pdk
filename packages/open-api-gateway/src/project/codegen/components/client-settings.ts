/*********************************************************************************************************************
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

 Licensed under the Apache License, Version 2.0 (the "License").
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ******************************************************************************************************************** */

import * as path from "path";
import { getLogger } from "log4js";
import { Component, JsonFile, Project } from "projen";
import { tryReadFileSync } from "projen/lib/util";
import {
  ClientLanguageConfig,
  DocumentationFormatConfig,
} from "../../client-config";
import { ClientLanguage, DocumentationFormat } from "../../languages";

// initialize logger
const logger = getLogger();

/**
 * Configuration for the ClientConfig component
 */
export interface ClientSettingsOptions {
  /**
   * The name of the client config file.
   * @default ".client-settings.json"
   */
  readonly clientSettingsFilename?: string;

  /**
   * The path to the generated code dir.
   */
  readonly generatedCodeDir: string;

  /**
   * Client languages set at current run.
   */
  readonly clientLanguages: ClientLanguage[];

  /**
   * Default client language.
   */
  readonly defaultClientLanguage: ClientLanguage;

  /**
   * Documentation formats set at current run.
   */
  readonly documentationFormats: DocumentationFormat[];

  /**
   * Indicates whether the OpenApiSpec has changed.
   */
  readonly specChanged: boolean;

  /**
   * Indicates whether the user set the forceGenerate flag.
   */
  readonly forceGenerateCodeAndDocs: boolean;
}

/**
 * Component for parsing the yaml OpenAPI spec as a single json object, resolving references etc.
 */
export class ClientSettings extends Component {
  public readonly clientSettingsPath: string;

  public readonly clientSettingsFilename: string;

  public readonly clientLanguageConfigs: ClientLanguageConfig[];
  public readonly documentationFormatConfigs: DocumentationFormatConfig[];

  private readonly options: ClientSettingsOptions;

  constructor(project: Project, options: ClientSettingsOptions) {
    super(project);

    this.options = options;

    this.clientSettingsFilename =
      options.clientSettingsFilename ?? ".client-settings.json";
    this.clientSettingsPath = path.join(
      options.generatedCodeDir,
      this.clientSettingsFilename
    );

    // load previously generated client config
    const clientSettingsPathAbs = path.join(
      project.outdir,
      this.clientSettingsPath
    );
    logger.trace(`Reading client settings from ${clientSettingsPathAbs}`);
    const existingClientConfig = tryReadFileSync(clientSettingsPathAbs);

    const prevClientLanguages = new Set<ClientLanguage>();
    const prevDocFormats = new Set<DocumentationFormat>();

    if (existingClientConfig) {
      const parsedClientConfig = JSON.parse(existingClientConfig);

      // previously generated client settings
      parsedClientConfig.clientLanguages.map((l: ClientLanguage) =>
        prevClientLanguages.add(l)
      );
      parsedClientConfig.documentationFormats.map((d: DocumentationFormat) =>
        prevDocFormats.add(d)
      );
    }

    this.clientLanguageConfigs = options.clientLanguages.map(
      (clientLanguage) => ({
        clientLanguage,
        isDefault: clientLanguage === options.defaultClientLanguage,
        generate:
          options.specChanged ||
          options.forceGenerateCodeAndDocs ||
          !prevClientLanguages.has(clientLanguage),
      })
    );

    this.documentationFormatConfigs = options.documentationFormats.map(
      (documentationFormat) => ({
        documentationFormat,
        generate:
          options.specChanged ||
          options.forceGenerateCodeAndDocs ||
          !prevDocFormats.has(documentationFormat),
      })
    );
  }

  synthesize(): void {
    // store the client config
    // this helps to optimize to re-generate projects when new language/doc format is
    // introduced while the spec doesn't change
    logger.trace(`Generating settings file to ${this.clientSettingsPath}`);
    const clientSettingsJsonFile = new JsonFile(
      this.project,
      this.clientSettingsPath,
      {
        obj: {
          clientLanguages: this.options.clientLanguages,
          documentationFormats: this.options.documentationFormats,
        },
        marker: true,
        editGitignore: false,
      }
    );
    logger.debug(
      `Generated settings file to ${clientSettingsJsonFile.absolutePath}`
    );
  }
}
