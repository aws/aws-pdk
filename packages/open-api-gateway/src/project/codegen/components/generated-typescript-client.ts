// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as path from "path";
import * as fs from "fs-extra";
import { Component, Project } from "projen";
import { exec } from "projen/lib/util";

/**
 * Configuration for the GeneratedTypescriptClient component
 */
export interface GeneratedTypescriptClientOptions {
  /**
   * Absolute path to the OpenAPI specification (spec.yaml)
   */
  readonly specPath: string;
}

/**
 * Generates the typescript client using OpenAPI Generator
 */
export class GeneratedTypescriptClient extends Component {
  private options: GeneratedTypescriptClientOptions;

  constructor(project: Project, options: GeneratedTypescriptClientOptions) {
    super(project);
    this.options = options;
  }

  synthesize() {
    super.synthesize();

    // Generate the typescript client
    exec(
      `./generate --spec-path ${this.options.specPath} --output-path ${this.project.outdir} --package-name ${this.project.name}`,
      {
        cwd: path.resolve(
          __dirname,
          "..",
          "..",
          "..",
          "..",
          "scripts",
          "generators",
          "typescript"
        ),
      }
    );

    // Write an index.ts which exposes the additional generated file OperationConfig.ts, which contains handler wrappers
    // and other generated code used by the construct.
    fs.writeFileSync(
      path.join(this.project.outdir, "src", "index.ts"),
      [
        "/* tslint:disable */",
        "/* eslint-disable */",
        "export * from './runtime';",
        "export * from './apis';",
        "export * from './models';",
        "export * from './apis/DefaultApi/OperationConfig';",
      ].join("\n")
    );
  }
}
