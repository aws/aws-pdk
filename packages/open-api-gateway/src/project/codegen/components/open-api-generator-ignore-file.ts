/*********************************************************************************************************************
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

 Licensed under the Apache License, Version 2.0 (the "License").
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ******************************************************************************************************************** */
import { IgnoreFile, Project } from "projen";

/**
 * Represents an .openapi-generator-ignore file
 * @see https://github.com/OpenAPITools/openapi-generator/blob/master/docs/customization.md#ignore-file-format
 */
export class OpenApiGeneratorIgnoreFile extends IgnoreFile {
  constructor(project: Project) {
    super(project, ".openapi-generator-ignore");

    // We should always use the projen .gitignore since projen manages the build, and therefore the ignored build
    // artifacts
    this.addPatterns(".gitignore");
  }
}
