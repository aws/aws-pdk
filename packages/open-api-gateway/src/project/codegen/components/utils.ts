/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as fs from "fs";
import * as path from "path";
import { exec } from "projen/lib/util";
import { ClientLanguage } from "../../languages";

/**
 * Enum for generator directories for non-client generators
 */
export enum NonClientGeneratorDirectory {
  DOCS = "docs",
}

/**
 * Generator directory for openapi generation containing templates, config etc.
 */
export type GeneratorDirectory = ClientLanguage | NonClientGeneratorDirectory;

/**
 * Options for generating client code or docs using OpenAPI Generator CLI
 */
export interface GenerationOptions {
  /**
   * The OpenAPI generator to use to generate the code/docs
   */
  readonly generator: string;
  /**
   * The directory to use for OpenAPI generation
   */
  readonly generatorDirectory: GeneratorDirectory;
  /**
   * The path of the OpenAPI spec to generate the client for
   */
  readonly specPath: string;
  /**
   * The directory in which the generated code should be output
   */
  readonly outputPath: string;
  /**
   * Additional properties to pass to the generate cli
   */
  readonly additionalProperties?: {
    [key: string]: string;
  };
}

const serializeProperties = (properties: { [key: string]: string }) =>
  Object.entries(properties)
    .map(([key, value]) => `${key}=${value}`)
    .join(",");

/**
 * Clean up any files that have already been generated by the openapi generator
 * @param generatedOutputPath output path for the generator
 */
const cleanPreviouslyGeneratedFiles = (generatedOutputPath: string) => {
  // OpenAPI generator writes a manifest called FILES which lists the files it generated.
  const openApiGeneratedFilesManifestPath = path.join(
    generatedOutputPath,
    ".openapi-generator",
    "FILES"
  );

  // If the manifest exists, delete the files it lists
  if (fs.existsSync(openApiGeneratedFilesManifestPath)) {
    const previouslyGeneratedFiles = new Set(
      fs
        .readFileSync(openApiGeneratedFilesManifestPath, { encoding: "utf-8" })
        .split("\n")
        .filter((x) => x)
    );
    previouslyGeneratedFiles.forEach((previouslyGeneratedFile) => {
      fs.unlinkSync(path.join(generatedOutputPath, previouslyGeneratedFile));
    });
  }
};

/**
 * Generate client code or docs by invoking the root generate script
 */
export const invokeOpenApiGenerator = (options: GenerationOptions) => {
  // Prior to generating, clean up any previously generated files so that we don't have stale generated code from
  // previous executions.
  cleanPreviouslyGeneratedFiles(options.outputPath);

  const additionalProperties = options.additionalProperties
    ? ` --additional-properties "${serializeProperties(
        options.additionalProperties
      )}"`
    : "";
  exec(
    `./generate --generator ${options.generator} --spec-path ${options.specPath} --output-path ${options.outputPath} --generator-dir ${options.generatorDirectory}${additionalProperties}`,
    {
      cwd: path.resolve(
        __dirname,
        "..",
        "..",
        "..",
        "..",
        "scripts",
        "generators"
      ),
    }
  );
};
