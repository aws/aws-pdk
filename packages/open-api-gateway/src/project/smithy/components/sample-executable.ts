/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as fs from "fs";
import * as path from "path";
import { Project, SampleFile, SampleFileOptions } from "projen";
import { getFilePermissions } from "projen/lib/util";

/**
 * A sample file that is executable
 */
export class SampleExecutable extends SampleFile {
  private readonly fullFilePath: string;

  constructor(project: Project, filePath: string, options: SampleFileOptions) {
    super(project, filePath, options);
    this.fullFilePath = path.join(this.project.outdir, filePath);
  }

  /**
   * @inheritDoc
   */
  public synthesize() {
    if (fs.existsSync(this.fullFilePath)) {
      return;
    }

    super.synthesize();

    fs.chmodSync(
      this.fullFilePath,
      getFilePermissions({
        executable: true,
        readonly: false,
      })
    );
  }
}
