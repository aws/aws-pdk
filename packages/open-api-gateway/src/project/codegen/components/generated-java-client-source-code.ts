/*********************************************************************************************************************
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

 Licensed under the Apache License, Version 2.0 (the "License").
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ******************************************************************************************************************** */

import { Component } from "projen";
import { JavaProject } from "projen/lib/java";
import { ClientLanguage } from "../../languages";
import { invokeOpenApiGenerator } from "./utils";

/**
 * Configuration for the GeneratedJavaClient component
 */
export interface GeneratedJavaClientSourceCodeOptions {
  /**
   * Absolute path to the OpenAPI specification (spec.yaml)
   */
  readonly specPath: string;
}

/**
 * Generates the java client using OpenAPI Generator
 */
export class GeneratedJavaClientSourceCode extends Component {
  private options: GeneratedJavaClientSourceCodeOptions;

  constructor(
    project: JavaProject,
    options: GeneratedJavaClientSourceCodeOptions
  ) {
    super(project);
    this.options = options;
  }

  /**
   * @inheritDoc
   */
  synthesize() {
    super.synthesize();

    const javaProject = this.project as JavaProject;
    const invokerPackage = `${javaProject.pom.groupId}.${javaProject.name}.client`;

    // Generate the java client
    invokeOpenApiGenerator({
      generator: "java",
      specPath: this.options.specPath,
      outputPath: this.project.outdir,
      generatorDirectory: ClientLanguage.JAVA,
      additionalProperties: {
        // TODO: Upgrade to openapi-generator 6.0.1 when released so that useSingleRequestParameter is honoured
        // https://github.com/OpenAPITools/openapi-generator/milestone/42
        // https://github.com/OpenAPITools/openapi-generator/pull/12580
        // This will be required for generating java lambda handler wrappers
        useSingleRequestParameter: "true",
        groupId: javaProject.pom.groupId,
        artifactId: javaProject.pom.artifactId,
        artifactVersion: javaProject.pom.version,
        invokerPackage,
        apiPackage: `${invokerPackage}.api`,
        modelPackage: `${invokerPackage}.model`,
        hideGenerationTimestamp: "true",
      },
    });
  }
}
