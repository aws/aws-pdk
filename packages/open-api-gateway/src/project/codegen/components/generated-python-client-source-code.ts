/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { getLogger } from "log4js";
import { Component } from "projen";
import { PythonProject } from "projen/lib/python";
import { ClientLanguage } from "../../languages";
import { invokeOpenApiGenerator } from "./utils";

const logger = getLogger();

/**
 * Configuration for the GeneratedPythonClient component
 */
export interface GeneratedPythonClientSourceCodeOptions {
  /**
   * Absolute path to the OpenAPI specification (spec.yaml)
   */
  readonly specPath: string;

  /**
   * Control if generator needs to be invoked
   */
  readonly invokeGenerator: boolean;
}

/**
 * Generates the python client using OpenAPI Generator
 */
export class GeneratedPythonClientSourceCode extends Component {
  private options: GeneratedPythonClientSourceCodeOptions;

  constructor(
    project: PythonProject,
    options: GeneratedPythonClientSourceCodeOptions
  ) {
    super(project);
    this.options = options;
  }

  /**
   * @inheritDoc
   */
  synthesize() {
    super.synthesize();

    if (this.options.invokeGenerator) {
      // Generate the python client
      logger.debug("Generating python client...");
      invokeOpenApiGenerator({
        generator: "python-experimental",
        specPath: this.options.specPath,
        outputPath: this.project.outdir,
        generatorDirectory: ClientLanguage.PYTHON,
        additionalProperties: {
          packageName: (this.project as PythonProject).moduleName,
          projectName: this.project.name,
        },
      });
    }
  }
}
