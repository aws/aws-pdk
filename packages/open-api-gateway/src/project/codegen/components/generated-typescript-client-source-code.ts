/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { writeFileSync } from "fs-extra";
import { getLogger } from "log4js";
import { Component } from "projen";
import { TypeScriptProject } from "projen/lib/typescript";
import { ClientLanguage } from "../../languages";
import { invokeOpenApiGenerator } from "./utils";

const logger = getLogger();

/**
 * Configuration for the GeneratedTypescriptClient component
 */
export interface GeneratedTypescriptClientSourceCodeOptions {
  /**
   * Absolute path to the OpenAPI specification (spec.yaml)
   */
  readonly specPath: string;

  /**
   * Control if generator needs to be invoked
   */
  readonly invokeGenerator: boolean;
}

/**
 * Generates the typescript client using OpenAPI Generator
 */
export class GeneratedTypescriptClientSourceCode extends Component {
  private options: GeneratedTypescriptClientSourceCodeOptions;

  constructor(
    project: TypeScriptProject,
    options: GeneratedTypescriptClientSourceCodeOptions
  ) {
    super(project);
    this.options = options;
  }

  /**
   * @inheritDoc
   */
  synthesize() {
    super.synthesize();

    if (this.options.invokeGenerator) {
      // Generate the typescript client
      logger.debug("Generating typescript client...");
      invokeOpenApiGenerator({
        generator: "typescript-fetch",
        specPath: this.options.specPath,
        outputPath: this.project.outdir,
        generatorDirectory: ClientLanguage.TYPESCRIPT,
        additionalProperties: {
          npmName: (this.project as TypeScriptProject).package.packageName,
          typescriptThreePlus: "true",
          useSingleParameter: "true",
          supportsES6: "true",
        },
      });

      // Write an index.ts which exposes the additional generated file OperationConfig.ts, which contains handler wrappers
      // and other generated code used by the construct.
      writeFileSync(
        path.join(this.project.outdir, "src", "index.ts"),
        [
          "/* tslint:disable */",
          "/* eslint-disable */",
          "export * from './runtime';",
          "export * from './apis';",
          "export * from './models';",
          "export * from './apis/DefaultApi/OperationConfig';",
        ].join("\n")
      );
    }
  }
}
