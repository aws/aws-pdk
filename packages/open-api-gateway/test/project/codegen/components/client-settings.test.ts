/*********************************************************************************************************************
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

 Licensed under the Apache License, Version 2.0 (the "License").
 You may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ******************************************************************************************************************** */
import * as fs from "fs-extra";
import { Project } from "projen";
import { directorySnapshot } from "projen/lib/util/synth";
import { ClientSettings } from "../../../../src/project/codegen/components/client-settings";
import {
  ClientLanguage,
  DocumentationFormat,
} from "../../../../src/project/languages";

const synthClientSettings = (
  project: Project,
  clientSettingsFilename?: string
) => {
  try {
    new ClientSettings(project, {
      clientSettingsFilename,
      generatedCodeDir: "generated",
      clientLanguages: [ClientLanguage.TYPESCRIPT],
      defaultClientLanguage: ClientLanguage.TYPESCRIPT,
      documentationFormats: [DocumentationFormat.HTML2],
      specChanged: false,
      forceGenerateCodeAndDocs: false,
    });
    project.synth();
    return directorySnapshot(project.outdir);
  } finally {
    fs.removeSync(project.outdir);
  }
};

describe("Client Settings Unit Tests", () => {
  it("Default client-settings file gets generated", () => {
    const project = new Project({
      name: "test",
    });
    expect(synthClientSettings(project)).toMatchSnapshot();
  });

  it("Custom-name client-settings file gets generated", () => {
    const project = new Project({
      name: "test",
    });
    expect(
      synthClientSettings(project, "custom-client-settings.json")
    ).toMatchSnapshot();
  });
});
