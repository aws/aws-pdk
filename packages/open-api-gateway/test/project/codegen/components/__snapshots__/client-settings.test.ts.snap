// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client Settings Unit Tests Custom-name client-settings file gets generated 1`] = `
Object {
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
generated/custom-client-settings.json
",
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "generated/custom-client-settings.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "clientLanguages": Array [
      "typescript",
    ],
    "documentationFormats": Array [
      "html2",
    ],
  },
}
`;

exports[`Client Settings Unit Tests Default client-settings file gets generated 1`] = `
Object {
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
generated/.client-settings.json
",
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "generated/.client-settings.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "clientLanguages": Array [
      "typescript",
    ],
    "documentationFormats": Array [
      "html2",
    ],
  },
}
`;
